File: tool/remaining_courses.py
#!/usr/bin/env python3
"""Print the "<language>/<course>" jobs that still need authoring.

A course counts as done only if its file exists and passes validation, so an
interrupted generation run can be resumed without redoing good work.

Usage:
    python3 tool/remaining_courses.py                 # every language
    python3 tool/remaining_courses.py tamil kannada   # only these
    python3 tool/remaining_courses.py --json          # as a JSON array
"""
import json
import pathlib
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from validate_courses import COURSES_DIR, Report, check_single_file  # noqa: E402


def main():
    args = [a for a in sys.argv[1:] if not a.startswith("--")]
    as_json = "--json" in sys.argv[1:]

    languages = args or sorted(p.name for p in COURSES_DIR.iterdir() if p.is_dir())
    remaining = []
    for language in languages:
        manifest = json.loads((COURSES_DIR / language / "manifest.json").read_text())
        for course in manifest["courses"]:
            path = COURSES_DIR / language / f"{course['id']}.json"
            report = Report()
            if path.exists():
                # check_single_file prints a summary line; silence it here.
                stdout, sys.stdout = sys.stdout, open("/dev/null", "w")
                try:
                    check_single_file(path, report)
                finally:
                    sys.stdout.close()
                    sys.stdout = stdout
                if not report.errors:
                    continue
            remaining.append(f"{language}/{course['id']}")

    if as_json:
        print(json.dumps(remaining))
    else:
        for job in remaining:
            print(job)
        print(f"\n{len(remaining)} course files remaining", file=sys.stderr)


if __name__ == "__main__":
    main()

File: tool/remaining_courses.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Print the "<language>/<course>" jobs that still need authoring.
#
# A course counts as done only if its file exists and passes validation, so an
# interrupted generation run can be resumed without redoing good work.
#
# Usage:
#   ruby tool/remaining_courses.rb                  # every language
#   ruby tool/remaining_courses.rb tamil kannada    # only these
#   ruby tool/remaining_courses.rb --json           # as a JSON array

require 'stringio'

require_relative 'courses'
require_relative 'validate_courses'

module RemainingCourses
  module_function

  def main(argv)
    as_json = argv.include?('--json')
    named = argv.reject { |arg| arg.start_with?('--') }

    remaining = Courses.languages(named).flat_map do |language|
      manifest = Courses.read_json(Courses::COURSES_DIR / language / 'manifest.json')
      manifest['courses'].filter_map do |course|
        path = Courses::COURSES_DIR / language / "#{course['id']}.json"
        next "#{language}/#{course['id']}" unless path.exist?

        report = Courses::Report.new
        # check_single_file prints a summary line; swallow it here.
        original = $stdout
        $stdout = StringIO.new
        begin
          ValidateCourses.check_single_file(path, report)
        ensure
          $stdout = original
        end
        report.ok? ? nil : "#{language}/#{course['id']}"
      end
    end

    if as_json
      puts JSON.generate(remaining)
    else
      remaining.each { |job| puts job }
      warn "\n#{remaining.size} course files remaining"
    end
    0
  end
end

exit RemainingCourses.main(ARGV) if $PROGRAM_NAME == __FILE__
